keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md

```diff
@@ -136,47 +135,49 @@ incorrectly or objects being garbage collected mistakenly.
     complete list of available resources. To fix the problems mentioned, it's
     necessary that discovery exactly matches the capability of the system. So,
-    we will use the storage version objects to reconstruct a merged discovery
+    we will use the remote [aggregate-discovery](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3352-aggregated-discovery/README.md) calls to peer apiserver to reconstruct a merged discovery
     document and serve that in all apiservers.
 
-Why so much work?
-* Note that merely serving 503s at the right times does not solve the problem,
-  for two reasons: controllers might get an incomplete discovery and therefore
-  not ask about all the correct resources; and when they get 503 responses,
-  although the controller can avoid doing something destructive, it also can't
-  make progress and is stuck for the duration of the upgrade.
-* Likewise proxying but not merging the discovery document, or merging the
-  discovery document but serving 503s instead of proxying, doesn't fix the
-  problem completely. We need both safety against destructive actions and the
-  ability for controllers to proceed and not block.
+    Why so much work?
+    * Note that merely serving 503s at the right times does not solve the problem,
+      for two reasons: controllers might get an incomplete discovery and therefore
+      not ask about all the correct resources; and when they get 503 responses,
+      although the controller can avoid doing something destructive, it also can't
+      make progress and is stuck for the duration of the upgrade.
+    * Likewise proxying but not merging the discovery document, or merging the
+      discovery document but serving 503s instead of proxying, doesn't fix the
+      problem completely. We need both safety against destructive actions and the
+      ability for controllers to proceed and not block.
+
+* Version Aware Proxying:
+
+  - With [compatibility versions](https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/4330-compatibility-versions/README.md) information available for kubernetes components, we can leverage that to perform version aware proxying between API servers, and ensure that a request is proxied to an eligible apiserver given that it was generated by a client compatible with the apiserver/s present in a cluster. Version aware proxying lets us ensure that requests from clients (possibly) expecting to use a feature introduced in later K8s versions,  are not incorrectly routed to apiservers at older versions
+
+  - We will add a new header in client requests to specify the compatibility version of the client 
+
+  - We will publish [min-compatibility version](https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/4330-compatibility-versions/README.md#--min-compatibility-version), binary-version of an apiserver as annotations in the apiserver-identity lease
```

Comment:
sweet! I don't see any section that states how we'll be using these annotations, but +1.

Reply:
Moved it to https://github.com/richabanker/enhancements/pull/1 to keep track of all compat-version-aware-proxy changes separately